src/macros.jl

```diff
@@ -166,3 +166,125 @@ end
 @inline (self::SkipOptimize{F})(a) where {F} = self.f(value(a))
 @inline (self::SkipOptimize{F})(a, b) where {F} = self.f(value(a), value(b))
 @inline (self::SkipOptimize{F})(a, b, c) where {F} = self.f(value(a), value(b), value(c))
+
+"""
+    f(x) = dot(x, x)
+    f(x::ReverseDiff.TrackedVector) = ReverseDiff.track(f, x)
+    ReverseDiff.@grad function f(x)
+        xv = ReverseDiff.value(x)
+        return dot(xv, xv), Δ -> (Δ * 2 * xv,)
+    end
+The `@grad` macro provides a way for the users to define custom adjoints for single-output functions wrt to their input numbers or arrays.
+"""
+macro grad(expr)
+    d = MacroTools.splitdef(expr)
+    f = d[:name]
+    closure = gensym(f)
+    d[:name] = closure
+    closure_ex = MacroTools.combinedef(d)
+
+    @gensym tp output_value output back args kwargs
+    args_ex = getargs_expr(d[:args])
+    kwargs_ex = getkwargs_expr(d[:kwargs])
+    return quote
+        function $ReverseDiff.track(::typeof($f), $(d[:args]...); $(d[:kwargs]...)) where {$(d[:whereparams]...),}
+            $closure_ex
+            $args = $args_ex
+            $kwargs = $kwargs_ex
+            $tp = $ReverseDiff.tape($args...)
+            $output_value, $back = $closure($args...; $kwargs...)
+            $output = $ReverseDiff.track($output_value, $tp)
+            $ReverseDiff.record!(
+                $tp,
+                $ReverseDiff.SpecialInstruction,
+                $f,
+                $args,
+                $output,
+                ($back, $closure, $kwargs),
+            )
+            return $output
+        end
+
+        if !hasmethod(
+            $ReverseDiff.special_reverse_exec!,
+            Tuple{$ReverseDiff.SpecialInstruction{typeof($f)}},
+        )
+            @noinline function $ReverseDiff.special_reverse_exec!(instruction::$ReverseDiff.SpecialInstruction{typeof($f)})
+                output = instruction.output
+                input = instruction.input
+                back = instruction.cache[1]
+                input_derivs = back($ReverseDiff.deriv(output))
+                @assert input_derivs isa Tuple
+                $ReverseDiff.add_to_deriv!.(input, input_derivs)
+                $ReverseDiff.unseed!(output)
+                return nothing
+            end
+        end
+
+        if !hasmethod(
+            $ReverseDiff.special_forward_exec!,
+            Tuple{$ReverseDiff.SpecialInstruction{typeof($f)}},
+        )
+            @noinline function $ReverseDiff.special_forward_exec!(instruction::$ReverseDiff.SpecialInstruction{typeof($f)})
+                output, input = instruction.output, instruction.input
+                $ReverseDiff.pull_value!.(input)
+                pullback = instruction.cache[2]
+                kwargs = instruction.cache[3]
+                out_value = pullback(input...; kwargs...)[1]
+                $ReverseDiff.value!(output, out_value)
+                return nothing
+            end
+        end
+    end |> esc
```

Comment:
is this necessary?

Reply:
Well I guess I can just `esc` the closure, argument types and where parameters but I just used `gensym` and `esc` all the way.

Reply:
oh yes, it's safe with gensym anyways, I see.